src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,404 @@
+
+public:
+  symbol_factoryt(
+    code_blockt &_init_code,
+    bool _assume_non_null,
+    symbol_tablet &_symbol_table,
+    input_recordert &input_recorder):
+      init_code(_init_code),
+      assume_non_null(_assume_non_null),
+      symbol_table(_symbol_table),
+      input_recorder(input_recorder),
+      ns(_symbol_table)
+  {}
+
+  exprt allocate_object(
+    const exprt &,
+    const typet &,
+    const source_locationt &,
+    const bool);
+
+  void gen_nondet_init(
+    const exprt &expr,
+    const source_locationt &loc);
+};
+
+/*******************************************************************\
+
+Function: symbol_factoryt::allocate_object
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const source_locationt &loc,
+  const bool static_lifetime)
+{
+  // Add the following code to init_code:
+  //   <allocate_type> tmp$<temporary_counter>;
+  //   <target_expr> = &tmp$<temporary_counter>
+  // Store the following code in the input recorder:
+  //   INPUT("tmp$<temporary_counter>", tmp$<temporary_counter>);
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  symbolt &aux_symbol=
+    c_declare_new_tmp_symbol(init_code, symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  expr -
+  loc -
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  const exprt &expr,
+  const source_locationt &loc)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+    std::string init_done_label_str;
+
+    if(!assume_non_null)
+    {
+      // Add the following code to init_code:
+      //   IF NONDET(_Bool) != FALSE THEN GOTO <set_null_label_str>
+      //   <code generated by allocate_object()>
+      //   <code generated by gen_nondet_init()>
+      //   GOTO <init_done_label_str>
+      //   <set_null_label_str>: <expr> = ((<pointer_type>)NULL);
+      //   <init_done_label_str>: SKIP
```

Comment:
I've done this. I had to make the auxiliary variables static, otherwise they were going out of scope at the end of the else clause, being declared dead, and CBMC didn't produce the right test suite.